web/controllers/v1/relay_group_membership_controller.ex

```diff
@@ -12,27 +12,36 @@ defmodule Cog.V1.RelayGroupMembershipController do
     render(conn, Cog.V1.RelayGroupView, "relays.json", relay_group: relay_group)
   end
 
+  def manage_membership(conn, %{"id" => id, "relays" => member_spec}),
+  do: manage_association(conn, :relay, %{"id" => id, "members" => member_spec})
+
+  def manage_assignment(conn, %{"id" => id, "bundles" => member_spec}),
+  do: manage_association(conn, :bundle, %{"id" => id, "members" => member_spec})
+
   # Manage membership of a relay group. Adds and deletes can be submitted and
   # processed in a single request.
-  def manage_membership(conn, %{"id" => id, "relays" => member_spec}) do
+  def manage_association(conn, type, %{"id" => id, "members" => member_spec}) do
     result = Repo.transaction(fn() ->
       group = Repo.get!(RelayGroup, id)
 
-      to_add = Enum.reduce(Map.get(member_spec, "add", []),
-                           %{found: [], not_found: []}, &lookup_or_fail/2)
-      to_remove = Enum.reduce(Map.get(member_spec, "remove", []),
-                           %{found: [], not_found: []}, &lookup_or_fail/2)
+      to_add = Enum.reduce(Map.get(member_spec, "add", []), %{found: [], not_found: []},
```

Comment:
I don't like how `lookup_or_fail/3` embeds logic from this reduce call in it, making _this_ code a little less transparent.

A cleaner approach would be to encapsulate the entire reduce call into a function of its own, passing in the `type`, and the list of IDs. Doing something with `Enum.group_by/2` would be better still; just have the new version of `lookup_or_fail` return `:found` or `:not_found` and you get a bunch of stuff for free :smile: